cli/src/nativeMain/kotlin/com/dropbox/differ/cli/Main.kt

```diff
@@ -1,40 +1,126 @@
 import kotlinx.cli.ArgParser
 import kotlinx.cli.ArgType
 import kotlinx.cli.default
-import kotlin.system.exitProcess
 
 fun main(args: Array<String>) {
   val parser = ArgParser("differ")
-  val threshold by parser.option(
-    ArgType.Double,
-    fullName = "threshold",
-    shortName = "t",
-    description = "Percent of pixels that can differ while still resulting in a success."
-  ).default(0.001)
+  val verbose by parser
+    .option(
+      ArgType.Boolean,
+      fullName = "verbose",
+      description = "Print verbose comparison output."
+    )
+    .default(false)
+  val threshold by parser
+    .option(
+      ArgType.Double,
+      fullName = "threshold",
+      shortName = "t",
+      description = "Percent of pixels that can differ while still resulting in a success."
+    )
+    .default(0.001)
+  val delta by parser
+    .option(
+      ArgType.Double,
+      fullName = "delta",
+      shortName = "d",
+      description = "Max distance colors can be in 4 dimensional space without triggering a difference."
+    )
+    .default(0.08)
+  val hShift by parser
+    .option(
+      ArgType.Int,
+      fullName = "h-shift",
+      description = "Acceptable horizontal shift of pixel.",
+    )
+    .default(0)
+  val vShift by parser
+    .option(
+      ArgType.Int,
+      fullName = "v-shift",
+      description = "Acceptable vertical shift of pixel.",
+    )
+    .default(0)
+  val outputFile by parser
+    .option(
+      ArgType.String,
+      fullName = "output",
+      shortName = "o",
+      description = "Path at which the output file should be written."
+    )
+
   val leftPath by parser.argument(ArgType.String, "First Image path", "Path to the first image for comparison")
   val rightPath by parser.argument(ArgType.String, "Second Image path", "Path to the second image for comparison")
 
   parser.parse(args)
 
-  val comparator = SimpleImageComparator()
-  val mask = withImage(leftPath) { leftImage ->
+  val startTime = getTimeMicros()
+  val comparator = SimpleImageComparator(
+    maxDistance = delta.toFloat(),
+    hShift = hShift,
+    vShift = vShift,
+  )
+  var mask: Mask
+
+  val result = withImage(leftPath) { leftImage ->
+    outputFile?.let {
+      writePPMFile(it, leftImage)
+    }
     withImage(rightPath) { rightImage ->
-      val mask = Mask(leftImage.width, leftImage.height)
+      mask = Mask(leftImage.width, leftImage.height)
       comparator.compare(leftImage, rightImage, mask)
-      mask
     }
   }
 
-  if (mask.difference > threshold) {
-    println("Image comparison failed with difference: ${mask.difference}")
-    exitProcess(1)
-  } else {
-    println("Image comparison succeeded.")
-    exitProcess(0)
+  val diff = result.pixelDifferences.toFloat() / result.pixelCount
+  val pass = diff <= threshold
+
+  if (verbose) {
+    val status = when {
+      result.pixelDifferences == 0 -> "identical"
+      pass -> "similar"
+      else -> "different"
+    }
+    println("Images are $status")
```

Comment:
Not a request for this PR, but it might be nice to have some improved logging in this.  Our CI console output are huge and difficult to follow.

Reply:
This is simply when the `--verbose` flag is provided, and it based directly on the existing blink-diff output.  I wouldn't expect it to be used in our CI pipeline, but we can expand this if we find more useful stuff.